pkg/credentialprovider/apis/criauth/types.go

```diff
@@ -0,0 +1,159 @@
+	// plugin response does not return an expiration duration, then credentials will not be cached by
+	// the kubelet.
+	DefaultCacheDuration metav1.Duration
+
+	// extraArgs specifies arguments to be passed to the plugin via environment variables.
+	// +optional
+	ExtraArgs []PluginArg
+}
+
+// PluginArg is used for passing extra arguments to the exec credential plugin binary.
+// Exec plugins expect the arguments to be passed in the format "key=value".
+type PluginArg struct {
+	Name  string
+	Value string
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CRIAuthPluginRequest includes the image that the kubelet requires authentication for.
+// Kubelet will pass this request object to the plugin via stdin.
+type CRIAuthPluginRequest struct {
+	metav1.TypeMeta
+
+	// image is the container image that is being pulled as part of the
+	// CRI auth plugin request. Plugins may optionally parse the image
+	// to extract any information required to fetch credentials.
+	Image string
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CRIAuthPluginResponse holds credentials that the kubelet should use for the specified
+// image provided in the original request. Kubelet will read the response from the plugin via stdout.
+type CRIAuthPluginResponse struct {
+	metav1.TypeMeta
+
+	// authEntries is a list of AuthEntry. Each AuthEntry should contain credentials that can be used
+	// for the provided image in CRIAuthPluginRequest. AuthEntry also contains metadata for the kubelet
+	// such as a match key and cache duration.
+	AuthEntries []AuthEntry
+}
+
+// AuthEntry contains credentials that can be used for the provided image in CRIAuthPluinRequest.
+// It also contains useful metadata for the kubelet such a match key and cache duration.
+type AuthEntry struct {
+	// matchKey is the match URL that credentials are valid for. The kubelet will use
+	// this value when looking up images in the auth keyring. This is also the cache key
+	// used by the kubelet.
+	//
+	// matchKey should be a URL string pattern which can optionally contain a port and path.
+	// Globs can be used in the hostname, but not in the port or the path. The kubelet will use
+	// the AuthConfig in this entry for any subsequent images matching against this match key while
+	// this entry is still cached.
+	//
+	// Examples of values:
+	//   - 123456789.dkr.ecr.us-east-1.amazonaws.com
+	//   - *.azurecr.io
+	//   - gcr.io
+	//   - *.*.registry.io
+	//   - registry.io:8080/path
+	MatchKey string
+
+	// cacheDuration indicates the duration the provided credentials should be cached for.
+	// The kubelet will use this field to set the in-memory cache duration for credentials
+	// in the AuthConfig. If nil, the kubelet will use defaultCacheDuration provided in
+	// CRIAuthPluginConfig. If set to 0, the kubelet will not cache the provided AuthConfig.
+	CacheDuration *metav1.Duration
```

Comment:
discussed during code review, I think we want to constrain the plugin's options for cacheKey to image-level, registry-level, and global, and make the plugin-specified cacheKey and cacheDuration apply to the whole plugin response, not individual auth entries.

I'd envision the provider flow for a single plugin to look like this:
```
kubelet calls Provide("example.com:8080/mypath/myimage")

	look in cache for "example.com:8080/mypath/myimage", return if cached response is present
	look in cache for "example.com:8080", return if cached response is present
	look in cache for "", return if cached response is present

If no cached results are present, invoke plugin with "example.com:8080/myimage"
	
Plugin responds with:
	cacheKey: "image" | "registry" | "global"
	cacheDuration: "0" | "60s" | ...
	authEntries: [...]

If cacheDuration is 0, do not cache results

Compute cache key based on the plugin response:
	image: "example.com:8080/mypath/myimage"
	registry: "example.com:8080"
	global: ""

Cache with the given key under for the plugin-specified duration
```